src/goto-instrument/contracts/dynamic-frames/dfcc_cfg_info.h

```diff
@@ -0,0 +1,293 @@
+#define CPROVER_GOTO_INSTRUMENT_CONTRACTS_DYNAMIC_FRAMES_DFCC_CFG_INFO_H
+
+#include <util/message.h>
+#include <util/namespace.h>
+#include <util/std_expr.h>
+
+#include <goto-programs/goto_program.h>
+
+#include "dfcc_loop_contract_mode.h"
+
+#include <map>
+#include <set>
+
+class dfcc_utilst;
+class dfcc_libraryt;
+class goto_modelt;
+class goto_functiont;
+
+/// \brief Describes a single loop for the purpose of DFCC loop contract
+/// instrumentation.
+class dfcc_loop_infot
+{
+public:
+  dfcc_loop_infot(
+    const std::size_t loop_id,
+    const std::size_t cbmc_loop_id,
+    const std::set<exprt> assigns,
+    const exprt invariant,
+    const exprt decreases,
+    const std::set<irep_idt> local,
+    const std::set<irep_idt> tracked,
+    const std::vector<std::size_t> inner_loops,
+    const std::vector<std::size_t> outer_loops,
+    const symbol_exprt write_set_var,
+    const symbol_exprt addr_of_write_set_var)
+    : loop_id(loop_id),
+      cbmc_loop_id(cbmc_loop_id),
+      assigns(assigns),
+      invariant(invariant),
+      decreases(decreases),
+      local(local),
+      tracked(tracked),
+      inner_loops(inner_loops),
+      outer_loops(outer_loops),
+      write_set_var(write_set_var),
+      addr_of_write_set_var(addr_of_write_set_var)
+  {
+  }
+
+  /// Prints a textual representation of the struct to \p out.
+  void output(std::ostream &out, namespacet &ns, const irep_idt &mode) const;
+
+  // Finds the instruction tagged as loop head and having the same loop
+  // identifier as this struct in the given program.
+  optionalt<goto_programt::instructiont::targett>
+  find_head(goto_programt &goto_program) const;
+
+  // Finds the last instruction tagged as loop latch hand having the same loop
+  // loop identifier as this struct in the given program.
+  optionalt<goto_programt::instructiont::targett>
+  find_latch(goto_programt &goto_program) const;
+
+  /// Loop identifier assigned by DFCC to this loop.
+  const std::size_t loop_id;
+
+  /// Loop identifier assigned to this loop by traditional CBMC loop numbering.
+  const std::size_t cbmc_loop_id;
+
+  /// Set of targets assigned by the loop, either user-provided or inferred.
+  const std::set<exprt> assigns;
+
+  /// Loop invariant expression.
+  const exprt invariant;
+
+  /// Decreases clause expression.
+  const exprt decreases;
+
+  /// Set of local identifiers locally DECL in loop instructions, excluding
+  /// identifiers declared in nested loops.
+  const std::set<irep_idt> local;
+
+  /// Subset of locals that must be tracked in the loop's write set.
+  /// A local must be tracked if it is dirty or might be assigned by one of
+  /// the inner loops of that loop.
+  const std::set<irep_idt> tracked;
+
+  /// Integer identifiers of inner loops of that loop.
+  const std::vector<std::size_t> inner_loops;
+
+  /// Integer identifier of the outer loop(s) if it exists.
+  const std::vector<std::size_t> outer_loops;
```

Comment:
Are these DFCC or CBMC loop identifiers? Also, the comment for `outer_loops` suggests that there can be at most one such loop? Should it be `optionalt` instead of `std::vector`?

Reply:
They are DFCC loop number.
No, there can be one or more outer loops.

Reply:
optionalt has caused some problems (spurious compiler errors about using uninitialised values)

Reply:
If this was just about compiler warnings, then we should silence these via a pragma. Let's not hack around it by using an unsuitable data structure.